### ISSUE TITLE
No compatible memory type

### ISSUE THREAD

**Issue by MASTER-OF-ORION** [yunyinghu/CVCT_Vulkan#2]
CVCT.cpp line 880...881 failed! :
[CRITICAL] No compatible memory type

in my case:
bufferTypeIndex == 0
stagingTypeIndex == 1

What is this? Do you keep a huge 3D Texture (line 4096^3) on the hard drive or what? Why did you need to use virtual memory, and even in such quantity? :-)

**Comment by yunyinghu** [comment]
Hey MoO!

I suspect that while setting up the local buffer holding all geometry data to use a flag that is not supported by your GPU. The staging buffer seems to be fine. What kind of GPU are you using? I'll set up a windows machine to test this issue tonight.

Regarding the huge amount of virtual memory; I'm virtual allocating four buffers for the following attributes:
- asset data (vertex data, textures - no mips are stored) (stored)
- asset descriptors (reconstructed depending on how the data is loaded)
- asset entries (stored)
- asset dependencies (stored)

When closing the application, the data, entries, and dependency buffers will be flushed to the HDD in this format:
		- header
		- cacheentries
		- data
		- dependencies

All three buffers have the same size of 0xFFFFFFFF, but honestly, except from the asset data buffer, they don't have any reason to be that big. In my next iteration, I'll scale them accordingly.

No voxel or texture 3D data will be stored. This iteration is currently caching nothing from the voxel data. The voxelization process is recomputed every frame, this is currently really inefficient and causes a lot of flickering of the indirect lighting. 

The iteration I'm currently working on will use conservative voxelization. So every frame checks if a new area has to be voxelized, and will only voxelize that area, while still conserving voxel data from previous frames. This will most likely make the flickering less apparent, and will speed up the voxelization process. Checking which area to voxelize will also be computed on the GPU using the indirectdraw feature in Vulkan, so there won't be any synchronisation between the host en device. To expand on this even further, the Vulkan API will release a proper multi GPU support soon. The conservative voxelization process seems perfect for offloading the work to a secondary GPU, if available.

I hope I answered your question regarding the 3D texture/voxels and the quantity of memory I'm allocating. I'll set up my windows machine tonight to fix the buffer issue.

Thanks for your interest and feedback!

**Comment by MASTER-OF-ORION** [comment]
My computer: video card: AMD Radeon HD 7800 serias, cpu - I7 3770

**Comment by yunyinghu** [comment]
My reason for using WinAPI is that I didn't have any plans on making it cross-platform. I was experimenting with a different way of managing my assets, and VirtualAlloc allowed me to allocate with a starting address. 

I am using atomic operations, but I think either the revoxelization or my algorithm to blend (mip-mapping) is causing my flickering. I remember the voxelization algorithm also having an issue with triangles that intersecting the same raster while rasterization. To mitigate this, I average the voxels.

Regarding the cascade algorithm, I used this as a main source:
http://fumufumu.q-games.com/archives/Cascaded_Voxel_Cone_Tracing_final_speaker_notes.pdf
https://research.nvidia.com/sites/default/files/publications/GIVoxels-pg2011-authors.pdf

The presentation compares the cascading process to the cascaded shadows, so most are interpreted from cascaded shadow maps. I couldn't find a paper that would describe the cascaded voxel algorithm directly. If you have any good resources, I would be happy if you could share them!

**Comment by MASTER-OF-ORION** [comment]
Thanks for links. 
I'm sorry, but I do not have documents for cascades ... I'm looking for them myself :-) The only thing I want to say is that I somewhere read that there is a method of better cascades - these are uses DirectX12 tiled resourses, but I did not really go into this...

**Comment by yunyinghu** [comment]
By tiled resources, you mean the sparse resources? Vulkan has something similar. On paper, this would seem perfect for storing the voxels, but you have to manage the paging of resources, which looks really interesting! But due to deadlines, I didn't have time to look into it as much as I would've loved to during development. 

I would love to discuss more with you about voxel cone tracing. Is there a way I can reach you? FB, skype?

**Comment by MASTER-OF-ORION** [comment]
Yes, indeed, in this world there are not so many people with whom one could talk about global illumination... Let's chat on skype, but .... better write your skype here, I don't want to show my skype to the whole internet )))))

**Comment by yunyinghu** [comment]
my Skype is valif.

**Comment by MASTER-OF-ORION** [comment]
Ohh...to many 'valif'-s in the list:-)
Write me better, my skype: darkness_ascendancy

2017-05-11 0:14 GMT+03:00 Yun Ying Hu <notifications@github.com>:
